receiver/prometheusreceiver/internal/transaction.go

```diff
@@ -159,6 +159,12 @@ func (tr *transaction) Commit() error {
 	if tr.useStartTimeMetric {
 		// AdjustStartTime - startTime has to be non-zero in this case.
 		if tr.metricBuilder.startTime == 0.0 {
+			// Unable to adjust start time because of missing start time metric
+			tr.logger.Info(
```

Comment:
Should this be a warning?

Reply:
I considered that initially, but since it's not the issue with collector or its config, but rather with the target applications, `Info` seems more appropriate?

i.e. collector itself performs correctly, it just informs us that some applications are misconfigured.

Reply:
I would consider it a warrning, as it informs about degraded behavior and there is a user action that can be taken to resolve it.

Reply:
We should not log on bad input, see https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#logging

Reply:
> We should not log on bad input

@tigrannajaryan could you suggest alternative solution then?
As described in #1921, we currently don't have any visibility in these metrics being dropped and why are they being dropped. Different people at Google ran into this issue in the last few months and spent hours / days debugging it.
Having a log message would help a lot.

As @serathius pointed above, in most cases this does require human intervention - either fixing the application, or change the collector config.

Reply:
Regardless of wether this is logged, there should probably be a call to `obsreport.EndMetricsReceiveOp(ctx, dataformat, 0, 0, errors.New(...))` here which would at mean these would show up as refused metrics and there would be traces reporting the error message.

Logging this doesn't seem so bad to me though as it is a pretty severe failure I think this code runs once per scrape rather than once per metric?

Reply:
Good point @james-bebbington !
Currently there is no error reported in https://github.com/open-telemetry/opentelemetry-collector/blob/447bceadb3f490dcf4b1aab500103bc086c4bf0b/receiver/prometheusreceiver/internal/transaction.go#L190-L192 for this case.

I do agree that this case is actually similar to https://github.com/open-telemetry/opentelemetry-collector/blob/447bceadb3f490dcf4b1aab500103bc086c4bf0b/receiver/prometheusreceiver/internal/transaction.go#L155-L156 where we do return an error.
I will update the PR with this change.

Leaving the log message for now as it is very helpful for debugging and as @james-bebbington said it only occurs once per scraping interval (per endpoint having the issue), i.e. not too frequent.

Reply:
FYI the error returned here is handled by Prometheus code, which will log it: https://github.com/prometheus/prometheus/blob/3240cf83f08e448e0b96a4a1f96c0e8b2d51cf61/scrape/scrape.go#L1074-L1077

Reply:
Actually the logger does contain Prometheus target: https://github.com/prometheus/prometheus/blob/3240cf83f08e448e0b96a4a1f96c0e8b2d51cf61/scrape/scrape.go#L259

so extra log message is redundant - will remove it from here.

Reply:
Done.

Reply:
LGTM

Reply:
Sorry for late reply, I was a away for a few days.

IMO, the right approach is to record the failures in an internal metric. The guidelines mention it:
>For such high-frequency events instead of logging consider adding an internal metric and increment it when the event happens.

I think obsreport.EndMetricsReceiveOp should do that.

If you want to also log the failure then I believe it is better to use logger.Debug() so that it is not enabled by default. Another alternate if it _must_ have more visibility is to log an error _once_ and clearly indicate in the error message that it will be only logged once. A third alternate is to use log rate limiting. zap logger seems to support it (I haven't tried it).